third_party/amd/lib/TritonAMDGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -115,6 +128,152 @@ struct LoadStoreConversionBase {
+            .Case<triton::BroadcastOp>([&](auto broadcastOp) {
+              return verifyNonNegativeExpr(broadcastOp.getSrc());
+            })
+            .Case<triton::ExpandDimsOp>([&](auto expandOp) {
+              return verifyNonNegativeExpr(expandOp.getSrc());
+            })
+            .Case<triton::SplatOp>([&](auto splatOp) {
+              return verifyNonNegativeExpr(splatOp.getSrc());
+            })
+            .Case<triton::MakeRangeOp>([&](auto makeRangeOp) {
+              return makeRangeOp.getStart() >= 0 && makeRangeOp.getEnd() >= 0;
+            })
+            .Case<arith::ConstantIntOp>(
+                [&](auto constIntOp) { return constIntOp.value() >= 0; })
+            .Case<arith::ConstantOp>([&](arith::ConstantOp constOp) {
+              Value val = constOp.getResult();
+              DenseIntElementsAttr constVal;
+              if (matchPattern(val, m_Constant(&constVal)) &&
+                  constVal.isSplat())
+                return constVal.getSplatValue<APInt>().isNonNegative();
+              return false;
+            })
+            .Case<triton::GetProgramIdOp>([&](auto pidOp) { return true; })
+            .Case<arith::MaxSIOp>([&](auto maxOp) {
+              // max(a,b) >= 0 iff a>=0 || b>=0
+              bool nnLhs = verifyNonNegativeExpr(maxOp.getLhs());
+              bool nnRhs = verifyNonNegativeExpr(maxOp.getRhs());
+              return nnLhs || nnRhs;
+            })
+            .Case<arith::RemSIOp>([&](auto remsiOp) {
+              // a % b >= 0 iff a>=0
+              return (verifyNonNegativeExpr(remsiOp.getLhs()));
+            })
+            .Case<arith::AddIOp, arith::MinSIOp, arith::MulIOp, arith::DivSIOp>(
+                // Generally speaking, a OP b >= 0  iff  a >= 0 && b >= 0 when
+                // OP != sub
+                [&](Operation *binOp) {
+                  bool nnLhs = verifyNonNegativeExpr(binOp->getOperand(0));
+                  bool nnRhs = verifyNonNegativeExpr(binOp->getOperand(1));
+                  return nnLhs && nnRhs;
+                })
+            .Default([&](Operation *op) {
+              // Base case 3: unknown operation
+              return false;
+            });
+    return nonNegative;
+  }
+
+  // Quick analysis on the Triton IR to decide if we can safely use
+  // buffer operations
+  bool canUseBufferOps(Value ptr) const {
+    // 1. Check if the pointer is uniform: i.e., if it comes from a scalar
+    // pointer(splatted) and non-uniform offset addition
+    DenseSet<triton::AddPtrOp> nonUniformUpdates;
+    SmallVector<Operation *> queue{ptr.getDefiningOp()};
+    while (!queue.empty()) {
+      Operation *curOp = queue.pop_back_val();
+      if (!curOp)
+        continue;
+      if (auto addPtrOp = dyn_cast<triton::AddPtrOp>(curOp))
+        if (isa<RankedTensorType>(addPtrOp.getPtr().getType()))
+          nonUniformUpdates.insert(addPtrOp);
+      for (Value operand : curOp->getOperands())
```

Comment:
Do we need to scan all operands? Just need to add the `getPtr().getDefiningOp()`?

Reply:
Yeah, I refactored this check, now it should be simpler and nicer to read. At the end of the day I am looking for a `addptr(splat(scalarPtr), offset)`, so I can simply pattern match this.